### PR TITLE
Add classes and methods to send email

### LINKED ISSUE #427: Send admins an email anytime a user account is created
### User stories

As an admin
When a user registers a new account
Then I receive an email notifying my of the new user account

### Notes

* Basic text email including username, full name, and email
* The scope of this ticket is to create the classes/functions/logic needed to build and send the email, but not actually send the email until we have #445 and #426 complete.

### PR DESCRIPTION
**Pull request by Elkrival**
closes #427

This pr adds the class and the template that is sent to an admin when a new user is signed up.

There are settings that need to be set up for AWS SES, but this should prepare most of the steps, except the credentials used for AWS because I do not have access to it.

The template will look like the image below. 

Screenshot 2023-10-25 at 1 51 25 PM

### CODE REVIEW

File: server/mailer/Mail.js [hunk 1]
```diff
@@ -0,0 +1,200 @@
+class Mail {
```
Comment (mikestone14) [inline server/mailer/Mail.js]: I think we should split this up into multiple classes to make it easy to extend to other types of mailers in the future. Something along the lines of:

```js
class BaseMailer {
  constructor({ to, from }) {
    this.to = to
    this.from = from
  }
  
  sendMail() {
    return this.mailService.sendEmail(this.emailParams).promise()
  }
  
  get mailService() {
    return new AWS.SES({ apiVersion: '2010-12-01' })
  }
  
  get emailParams() {
    return {
      Destination: {
        ToAddresses: this.to,
      },
      Source: this.from,
      Message: {
        Body: {
          Html: {
            Charset: 'UTF-8',
            Data: this.body,
          },
        },
        Subject: {
          Charset: 'UTF-8',
          Data: this.subject,
        },
      },
    }
  }
}
```

Then we can extend the Mail class for our specific needs:

```js
class RegistrationMailer extends BaseMailer {
  constructor(user) {
    super({ to: [process.env.ADMIN_EMAIL], from: process.env.EMAIL_SENDER })
    this.subject = 'New Registered User'
    this.user = user
  }
  
  get body() {
    // some template string
  }
}
```

Then we can call this as:

```js
const registrationMailer = new RegistrationMailer(user)
registrationMailer.sendMail()
```

Comment (Elkrival) [reply server/mailer/Mail.js]: [bd42052](https://github.com/AMP-SCZ/dpdash/pull/457/commits/bd42052b840fbf1079bbd0f09382a5568c3c9a20)

File: server/mailer/templates/registrationTemplate.ejs [hunk 2]
```diff
@@ -0,0 +1,345 @@
+`<!DOCTYPE html>
```
Comment (mikestone14) [inline server/mailer/templates/registrationTemplate.ejs]: Is this opening tick-mark intentional?

```suggestion
<!DOCTYPE html>
```

File: server/mailer/RegistrationMailer.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  get body() {
+    const registrationEmailTemplate = path.join(
+      __dirname,
+      'server',
+      'mailer',
+      'templates',
+      'registrationTemplate.ejs'
+    )
+
+    return renderFile(registrationEmailTemplate, user)
```
Comment (mikestone14) [inline server/mailer/RegistrationMailer.js]: do you need to pass the user in an object to access it in the template as "user"?

```suggestion
    return renderFile(registrationEmailTemplate, { user })
```

File: server/mailer/templates/registrationTemplate.ejs [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+          />
+          <path
+            d="M1760 118 c0 -8 20 -12 48 -11 25 1 5 -22 -21 -25 -39 -5 -33 -24 6
+  -20 28 3 32 7 35 34 3 29 1 31 -32 30 -20 -1 -36 -4 -36 -8z"
+          />
+        </g>
+      </svg>
+      <div>
+        <h1 style="font-weight: 300; text-transform: capitalize">
+          <%= user.name %> using <%= $user.mail %>
```
Comment (mikestone14) [inline server/mailer/templates/registrationTemplate.ejs]: why does one of the `user` objects get prefixed with `$` and the other doesn't?

File: server/mailer/templates/registrationTemplate.ejs [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+  0 -35 -2 -35 -32z"
+          />
+          <path
+            d="M1760 118 c0 -8 20 -12 48 -11 25 1 5 -22 -21 -25 -39 -5 -33 -24 6
+  -20 28 3 32 7 35 34 3 29 1 31 -32 30 -20 -1 -36 -4 -36 -8z"
+          />
+        </g>
+      </svg>
+      <div>
+        <h1 style="font-weight: 300; text-transform: capitalize">
```
Comment (mikestone14) [inline server/mailer/templates/registrationTemplate.ejs]: do we need all these styles and the svg? IMO a simple email conveying the information the receiver needs is more effective.

File: server/mailer/templates/registrationTemplate.ejs [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+            font-size: 1.1em;
+            color: rgba(0, 0, 0, 0.4);
+            margin-top: 10px;
+          "
+        >
+          is requesting access to dpdash
+        </h2>
+        <a
+          style="text-decoration: none"
+          href="https://staging.dpdash.com"
```
Comment (mikestone14) [inline server/mailer/templates/registrationTemplate.ejs]: is there a specific route we can do to? Also, should this be an env var? I imagine on prod we won't want to go to staging.dpdash.com.

File: server/mailer/templates/registrationTemplate.ejs (hunk 5 above)
Comment (Elkrival) [reply server/mailer/templates/registrationTemplate.ejs]: I didn't have any copy of what the email should say so I took the liberty of adding their logo to make it look and feel a bit more official.

File: server/mailer/RegistrationMailer.js (hunk 3 above)
Comment (Elkrival) [reply server/mailer/RegistrationMailer.js]: I don't think we need to wrap it in brackets, the user is already an object.

File: server/mailer/templates/registrationTemplate.ejs (hunk 5 above)
Comment (mikestone14) [reply server/mailer/templates/registrationTemplate.ejs]: let's go with the simplest approach for now. If they want to make it look better we can include Erica/Berfin to design something. This is an internal email, though, so it just needs the information.

File: server/mailer/BaseMailer.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    this.to = to
+    this.from = from
+  }
+
+  sendMail() {
+    return this.mailService.sendEmail(this.emailParams).promise()
+  }
+
+  get mailService() {
+    return new AWS.SES({ apiVersion: '2010-12-01' })
```
Comment (mikestone14) [inline server/mailer/BaseMailer.js:12]: is this the latest API version? I hope they have one from the last decade 😆

File: server/mailer/RegistrationMailer.js (hunk 3 above)
Comment (mikestone14) [reply server/mailer/RegistrationMailer.js]: but how does the template know the variable name?

File: server/mailer/BaseMailer.js (hunk 7 above)
Comment (Elkrival) [reply server/mailer/BaseMailer.js:12]: that's a good point, it's what they have listed in their documentation. 🤔 
https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/ses-examples-sending-email.html#:~:text=var%20sendPromise%20%3D%20new%20AWS.SES(%7BapiVersion%3A%20%272010%2D12%2D01%27%7D).sendTemplatedEmail(params).promise()%3B

File: server/mailer/templates/registrationTemplate.ejs (hunk 6 above)
Comment (Elkrival) [reply server/mailer/templates/registrationTemplate.ejs]: i've added the env var
[bc9cbcf](https://github.com/AMP-SCZ/dpdash/pull/457/commits/bc9cbcfb5650ebceec9074f06d4e04dcc47fc58f)

File: server/mailer/templates/registrationTemplate.ejs (hunk 5 above)
Comment (Elkrival) [reply server/mailer/templates/registrationTemplate.ejs]: 👍 
[bc9cbcf](https://github.com/AMP-SCZ/dpdash/pull/457/commits/bc9cbcfb5650ebceec9074f06d4e04dcc47fc58f)

File: server/mailer/templates/registrationTemplate.ejs (hunk 4 above)
Comment (Elkrival) [reply server/mailer/templates/registrationTemplate.ejs]: [bc9cbcf](https://github.com/AMP-SCZ/dpdash/pull/457/commits/bc9cbcfb5650ebceec9074f06d4e04dcc47fc58f)

File: server/mailer/RegistrationMailer.js (hunk 3 above)
Comment (Elkrival) [reply server/mailer/RegistrationMailer.js]: I've updated the template.
[bc9cbcf](https://github.com/AMP-SCZ/dpdash/pull/457/commits/bc9cbcfb5650ebceec9074f06d4e04dcc47fc58f)

File: server/mailer/templates/registrationTemplate.ejs (hunk 2 above)
Comment (Elkrival) [reply server/mailer/templates/registrationTemplate.ejs]: [bc9cbcf](https://github.com/AMP-SCZ/dpdash/pull/457/commits/bc9cbcfb5650ebceec9074f06d4e04dcc47fc58f)

File: server/mailer/templates/registrationTemplate.ejs [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+          font-size: 1.1em;
+          color: rgba(0, 0, 0, 0.4);
+          margin-top: 10px;
+        "
+      >
+        <h1><%= name %> using <%= mail %></h1>
+        <h2>is requesting access to dpdash</h2>
+        <a
+          style="text-decoration: none"
+          href="<%= process.env.HOME_URL %>"
```
Comment (mikestone14) [inline server/mailer/templates/registrationTemplate.ejs]: let's pass this in to the template like we do the user so we're not mixing where the data is coming from.

Comment (Elkrival) [reply server/mailer/templates/registrationTemplate.ejs]: [d7d81f3](https://github.com/AMP-SCZ/dpdash/pull/457/commits/d7d81f3f15d9919a2e74cb19a78f4a87d984c42c)

File: server/mailer/RegistrationMailer.js [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+    const registrationEmailTemplate = path.join(
+      __dirname,
+      'mailer',
+      'templates',
+      'registrationTemplate.ejs'
+    )
+
+    return renderFile(registrationEmailTemplate, {
+      ...user,
+      redirect: this.redirectUrl,
```
Comment (mikestone14) [inline server/mailer/RegistrationMailer.js:23]: this is probably fine, but if we were to have a `redirect` attribute on a user we'd be overriding it here and it also might cause confusion if `redirect` isn't an attribute of a user (which it isn't) but we're using it as one here. Might be worth not de-structuring and just passing `{ user, redirect: process.env.HOME_URL }` 🤷‍♂️

**Review by mikestone14 — APPROVED** [review]

**Pull request merged**